### ISSUE TITLE
Kafka source pod keep recreating for multi-partition kafka topic

### ISSUE THREAD

**Issue by bichitra95** [numaproj/numaflow#642]
**Describe the bug**

Kafka source pod fails to `Terminate` on default heartbeat and keeps recreating the pods when configured with Kafka topic which is having multiple partitions.

**To Reproduce**

*Local environment setup:*

* Installed Minikube locally
* Added Numaflow to the cluster from https://numaflow.numaproj.io/quick-start/#installation 
* Added helm chart of Kafka from here: https://bitnami.com/stack/kafka/helm
* Created a Kafka topic with the name `input-topic-partition` and the config looks like the below (having 2 Partitions).
<img width="1434" alt="Screenshot 2023-03-29 at 10 40 40 AM" src="https://user-images.githubusercontent.com/32828151/228434741-0478bbf7-870c-4d91-8f40-93575dcd7535.png">

*Pipeline:*

Installed a sample Kafka pipeline consisting of 3 vertices with the name `sample-kafka-pipeline`,
* Kafka source vertex to read the message from a Kafka topic
* UDF to do some processing over the message (set the expression to always true)
* Kafka sink to push message to a different Kafka topic
```apiVersion: numaflow.numaproj.io/v1alpha1
kind: Pipeline
metadata:
  name: sample-kafka-pipeline
spec:
  # Kafka Pipeline
  vertices:
    # Sources
    - name: input
      source:
        kafka:
          brokers:
            - 'local-kafka-0.local-kafka-headless.default.svc.cluster.local:9092'
          topic: input-topic-partition
          consumerGroup: numaflow-consumer-group

    # User defined functions
    - name: filter-messages
      udf:
        builtin:
          name: filter
          kwargs:
            expression: 1 < 100

    # Sinks
    - name: output-sink
      sink:
        kafka:
          brokers:
            - 'local-kafka-0.local-kafka-headless.default.svc.cluster.local:9092'
          topic: output-topic

  # The relationship between the vertices
  edges:
    - from: input
      to: filter-messages
    - from: filter-messages
      to: output-sink

```
**Steps to reproduce the behaviour:**
* After installing the pipeline in the cluster (in default namespace), add a few messages to the topic `input-topic-partition`.
* Wait for some time so that numaflow-controller will kill those vertices pods because of 10 heartbeats.
* Trigger a new message again and watch the pods in the cluster.
* Can be seen that multiple pods keep recreating with the name prefix `sample-kafka-pipeline-input-*` 

**Expected behaviour**
There should be one pod for the source as configured. but the pods keep creating whenever the controller tries to Terminate the pods and create those back.

**Screenshots**

Pods: 
<img width="1373" alt="Screenshot 2023-03-29 at 11 08 56 AM" src="https://user-images.githubusercontent.com/32828151/228436946-f1fafc34-960a-4506-99f4-ef073f530590.png">

Logs:

```kubectl logs sample-kafka-pipeline-input-0-jahju```
<img width="1373" alt="image" src="https://user-images.githubusercontent.com/32828151/228437421-9375d716-c733-4af8-bc59-43b59e515835.png">

```kubectl logs sample-kafka-pipeline-daemon-9cb96c99f-gwhrb```
<img width="1373" alt="image" src="https://user-images.githubusercontent.com/32828151/228437805-b2ab99a8-56d1-4e0f-9d37-e4957c10aecf.png">

**Environment (please complete the following information):**

- kubeletVersion: v1.25.3
- Minikube: v1.28.0
- Numaflow: v0.7.2

**Additional context**

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We often sort issues this way to know what to prioritize.

**Comment by vigith** [comment]
it looks like this is a problem due to auto-scaling. since this is a test env with very low TPS, the auto-scaling is causing this nuisance. Could you please try with auto-scaling turned off https://numaflow.numaproj.io/user-guide/autoscaling/#numaflow-autoscaling ?

**Comment by bichitra95** [comment]
Thanks @vigith, for looking at the issue. 

To add to your point, what I observed is that auto-scaling is working fine with all other vertices and also with this Kafka source vertex when it is configured with a Kafka topic of a single partition. Seems like auto-scaling is failing to scale down and up when the partition is more than one. 

Btw I will check to turn off auto-scaling for that vertex and come back with my observation.

**Comment by vigith** [comment]
We do use numaflow autoscaling on multi partitioned Kafka in production. Please let us know what behavior you see.

**Comment by bichitra95** [comment]
> it looks like this is a problem due to auto-scaling. since this is a test env with very low TPS, the auto-scaling is causing this nuisance. Could you please try with auto-scaling turned off https://numaflow.numaproj.io/user-guide/autoscaling/#numaflow-autoscaling ?

Tried turning off auto-scaling for the source pod only keeping everything else as before, works fine(recreation not happening). All other pods are Terminating and coming back as expected and the Kafka source pod is keep running.

Though we might require auto-scaling for kafka sources as well in production.

**Comment by ashwinidulams** [comment]
we do run with autoscaling in production with a multi partitioned topic. how many partitions do you have in your topic?
and whats the TPS in the topic?

**Comment by bichitra95** [comment]
I had set the partitions to 2 and TPS should not be the issue as I am testing this locally with just <5 messages moreover the issue occurs if the pods are ideal and hit 10 heartbeats.

**Comment by KeranYang** [comment]
### Investigation Notes
I use a k3d cluster to try to reproduce the issue.

When I specify my input kafka topic to have 2 partitions, when there is the no traffic, the source vertex **DOES NOT** scaled down to 0 pod. When 1 partition, it can scale down to 0.

I didn't see the pod re-creating case.

**Comment by KeranYang** [comment]
Hey @bichitra95 ,

Thank you for raising this issue.

As mentioned above, I was able to replicate the issue in which the Kafka source vertex is unable to scale down to 0 pods even when there are no input messages. This is due to an error in our pending Kafka message calculation, specifically when there are multiple Kafka input topic partitions but few messages, resulting in some partitions being empty. This miscalculation affects our auto-scaling feature.

I have addressed this issue by submitting a PR that fixes the bug. You can find more details about the fix in the PR: #666 

Unfortunately, I was unable to reproduce the problem of the pod continuously recreating itself. However, I believe it could be related to the bug I just fixed. We provide a nightly build for numaflow, so once my PR is merged, you can try it out to see if the recreating issue persists.

Please let me know if you require any additional assistance with this issue.

Thanks,
Keran Yang

**Comment by bichitra95** [comment]
Thanks, @KeranYang for the PR. I will try to test once the PR gets merged.
